src/node_i18n.cc

```diff
@@ -62,6 +72,306 @@ bool flag_icu_data_dir = false;
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  if (msg == nullptr || msg[0] == '\0')
+    msg = "Unspecified ICU Exception";
+  char buf[kStorageSize];
+  int len = snprintf(buf, sizeof(buf), "%s [%s]", msg, u_errorName(status));
+  Local<Value> e = Exception::Error(STRINGLEN(isolate, buf, len));
+  if (e.IsEmpty())
+    e = Local<Value>();
+  Local<Object> obj = e.As<Object>();
+  obj->Set(env->context(),
+           env->errno_string(),
+           Integer::New(isolate, status)).FromJust();
+  obj->Set(env->context(),
+           env->code_string(),
+           STRING(isolate, u_errorName(status))).FromJust();
+  env->isolate()->ThrowException(e);
+}
+
+// TODO(jasnell): This could potentially become a member of MaybeStackBuffer
+// at some point in the future. Care would need to be taken with the
+// MaybeStackBuffer<UChar> variant below.
+MaybeLocal<Object> AsBuffer(Isolate* isolate,
+                            MaybeStackBuffer<char>* buf,
+                            size_t len) {
+  if (buf->IsAllocated()) {
+    MaybeLocal<Object> ret = Buffer::New(isolate, buf->out(), len);
+    if (!ret.IsEmpty()) buf->Release();
+    return ret;
+  }
+  return Buffer::Copy(isolate, buf->out(), len);
+}
+
+MaybeLocal<Object> AsBuffer(Isolate* isolate,
+                            MaybeStackBuffer<UChar>* buf,
+                            size_t len) {
+  char* dst = reinterpret_cast<char*>(buf->out());
+  if (IsBigEndian())
+    SwapBytes16(dst, len * sizeof(UChar));
+  if (buf->IsAllocated()) {
+    MaybeLocal<Object> ret = Buffer::New(isolate, dst, len);
+    if (!ret.IsEmpty()) buf->Release();
+    return ret;
+  }
+  return Buffer::Copy(isolate, dst, len);
+}
+
+struct Converter {
+  Converter(Environment* env, const char* name)
+      : conv(nullptr) {
+    UErrorCode status = U_ZERO_ERROR;
+    conv = ucnv_open(name, &status);
+    CHECK(U_SUCCESS(status));
+  }
+
+  ~Converter() {
+    if (conv != nullptr)
+      ucnv_close(conv);
+  }
+
+  UConverter* conv;
+};
+
+// One-Shot Converters
+
+// Converts the Buffer from one named encoding to another.
+// args[0] is a string identifying the encoding we're converting to.
+// args[1] is a string identifying the encoding we're converting from.
+// args[2] must be a buffer instance
+void Convert(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[2]);
+
+  Utf8Value to_name(isolate, args[0]);
+  Utf8Value from_name(isolate, args[1]);
+
+  SPREAD_BUFFER_ARG(args[2], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  MaybeStackBuffer<char> buf;
+  uint32_t limit;
+  char* target;
+  const char* data = ts_obj_data;
+
+  Converter to(env, *to_name);
+  Converter from(env, *from_name);
+  if (to.conv == nullptr || from.conv == nullptr)
```

Comment:
Same question.  If the idea is that the caller should check that `converter.conv != nullptr` before continuing, then the CHECK in the constructor is probably wrong.